src/Transform/ONNX/ConstProp.cpp

```diff
@@ -217,6 +217,34 @@ struct ElementWiseBinaryOpImpl<ONNXMaxOp, T> {
   static T eval(T lhs, T rhs) { return std::max<T>(lhs, rhs); }
 };
 
+template <>
+struct ElementWiseBinaryOpImpl<ONNXModOp, int64_t, EnableNotBool<int64_t>> {
+  static int64_t eval(int64_t lhs, int64_t rhs) {
+    // The original calculation for mod
```

Comment:
I guess the whole computation here can be shorten to:
```
int64_t rem = remainder(lhs, rhs);
if ((rem != 0) && ((lhs < 0) ^ (rhs < 0)))
  return (rem + rhs);
return rem;
```

Reply:
Thanks for the review @tungld but there is a reason why I am using `lhs - floor(lhs / rhs) * rhs;` and not `remainder(lhs, rhs);`. Remainder is not always accurate in calculating the mod and I have a feeling that is the reason why mod is not working correctly. We use remainder in `Elementwise.cpp` but from testing we are not always getting the accurate results. I have calculated mod by hand and using google search engine and at times it differs from the results when using remainder. So to be safe I would rather do the following:

```
  // The original calculation for mod
    int mod = lhs - floor(lhs / rhs) * rhs;
    // Check if lhs is a multiple of rhs
    int64_t rem = remainder(lhs, rhs);

    // Handle the case when one of the int values are negative
    // If both int values are positive or multiples of each other, we can
    // calculate as normal
    if ((rem != 0) && ((lhs < 0) ^ (rhs < 0)))
      return (mod + rhs);
    return mod;
  }
};

```
The test below fails when using remainder... it is saying 5 mod 8 should be -3 but it actually should be 5. So this is another reason why I will not use remainder... the results are off. 

```
LC/llvm-project/build/bin/FileCheck /Users/meganhampton/zDLC/onnx-mlir/test/mlir/onnx/onnx_constprop.mlir
+ /Users/meganhampton/zDLC/onnx-mlir/build/Debug/bin/onnx-mlir-opt --shape-inference --constprop-onnx /Users/meganhampton/zDLC/onnx-mlir/test/mlir/onnx/onnx_constprop.mlir -split-input-file
+ /Users/meganhampton/zDLC/llvm-project/build/bin/FileCheck /Users/meganhampton/zDLC/onnx-mlir/test/mlir/onnx/onnx_constprop.mlir
/Users/meganhampton/zDLC/onnx-mlir/test/mlir/onnx/onnx_constprop.mlir:490:12: error: CHECK: expected string not found in input
 // CHECK: [[CONST:%.+]] = onnx.Constant dense<5> : tensor<i64>
           ^
<stdin>:335:49: note: scanning from here
 func.func @test_modulo_int_pos() -> tensor<i64> {
                                                ^
<stdin>:336:2: note: possible intended match here
 %0 = onnx.Constant dense<-3> : tensor<i64>
 ^

Input file: <stdin>
Check file: /Users/meganhampton/zDLC/onnx-mlir/test/mlir/onnx/onnx_constprop.mlir

-dump-input=help explains the following input dump.

Input was:
<<<<<<
             .
             .
             .
           330: } 
           331:  
           332:  
           333: // ----- 
           334: module { 
           335:  func.func @test_modulo_int_pos() -> tensor<i64> { 
check:490'0                                                     X~~ error: no match found
           336:  %0 = onnx.Constant dense<-3> : tensor<i64> 
check:490'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:490'1      ?                                           possible intended match
           337:  onnx.Return %0 : tensor<i64> 
check:490'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           338:  } 
check:490'0     ~~~
           339: } 
check:490'0     ~~
           340:  
check:490'0     ~
           341:  
check:490'0     ~
             .
             .
             .
>>>>>>

--
```

Reply:
OK, so it's because `remainder` works on float type. So it would be:
```
int64_t rem = lhs % rhs;
if ((rem != 0) && ((lhs < 0) ^ (rhs < 0)))
  return (rem + rhs);
return rem;

Reply:
@tungld Using ‘lhs%rhs’ does not work with eval. I have the error messages to prove it. I’m not sure what is wrong with using ‘floor’. It works accurately compared to remainder and all 6 of my lit tests pass. If we don’t have a reason why we shouldn’t use what I have, I’m not understanding. Thanks.

Reply:
Can you elaborate on `@tungld Using ‘lhs%rhs’ does not work with eval.`?

Reply:
I have used ‘%’ in eval and received an error message something like this operation can not be performed. I can send you the direct error message once I have my laptop.

Reply:
@tungld This was the error message:

```
/Users/meganhampton/zDLC/onnx-mlir/src/Transform/ONNX/ConstProp.cpp:230:44: error: invalid operands to binary expression ('double' and 'double')
  static T eval(T lhs, T rhs) { return lhs % rhs; }
```

But because you wanted me to switch to using two templates and the data type is specified it works. Before it was when I was using `T`. I made the updates, I think this is ready to be merged in now. Thanks